odoo/addons/base/tests/test_base.py

```diff
@@ -49,17 +73,324 @@ def test_04_literal_eval_forbidden(self):
+
+        safe_eval("time.strftime('%Y-%m-%d')", mode="exec", globals_dict={"time": time})
+
+    def test_14_deny_use_kwarg(self):
+        with self.assertRaisesRegex(SyntaxError, "unsafe node: variadic keyword arguments"):
+            safe_eval("def f(x, **kwargs): ...", mode="exec")
+
+        with self.assertRaisesRegex(SyntaxError, "unsafe node: variadic keyword arguments"):
+            safe_eval("lambda x, **kwargs: ...", mode="exec")
+
+    def test_15_deny_use_decorator(self):
+        with self.assertRaisesRegex(SyntaxError, "unsafe node: decorators"):
+            safe_eval(cleandoc(
+                """
+                @foo
+                def f(x):
+                    ...
+                """
+            ), mode="exec")
+
+    def test_16_deny_use_dunder_exception(self):
+        with self.assertRaisesRegex(SyntaxError, r"unsafe node: __type_checker \(dunder name\)"):
+            safe_eval(cleandoc(
+                """
+                try:
+                    ...
+                except Exception as __type_checker:
+                    ...
+                """
+            ), mode="exec")
+
+    def test_17_globals_dict_leak_builtins(self):
+        def set_locals(o):
+            frame = o.__traceback__.tb_frame
+            frame.f_locals['is_injected'] = True
+            frame.f_locals['__test__'] = 42
+
+        globals_dict = {"__bad_stuff": 0, "set_locals": set_locals}
+        locals_dict = {"__bad_stuff": 0}
+        safe_eval(cleandoc(
+        """
+        try:
+            1 / 0
+        except Exception as e:
+            set_locals(e)
+        """
+        ), globals_dict=globals_dict, locals_dict=locals_dict, mode="exec")
+        self.assertNotIn("__builtins__", globals_dict)
+        self.assertNotIn("__call_checker", globals_dict)
+        self.assertNotIn("__type_checker", globals_dict)
+        self.assertNotIn("__SafeWrapper", globals_dict)
+        self.assertNotIn("__bad_stuff", globals_dict)
+        self.assertNotIn("__bad_stuff", locals_dict)
+        self.assertNotIn("__test__", locals_dict)
+        self.assertTrue(locals_dict["is_injected"])
+
+    @patch("odoo.tools.safe_eval.SafeWrapper")
+    def test_18_bare_modules_becomes_wraps(self, mock):
+        # If the `type_checker` detect a bare module, and that the module is in the white-list
+        # then it should be wrapped in a SafeWrapper
+        c = ast.unparse(CodeChecker().visit(ast.parse("datetime.datetime")))
+        glb = _prepare_environment()
+        glb["__SafeWrapper"] = mock
+
+        unsafe_eval(compile(c, "", "exec"), glb, {"datetime": datetime})
+        mock.assert_called_once()
+
+    def test_19_wrap_on_decorator(self):
+        @allow_type
+        class Foo:
+            def __init__(self):
+                ...
+
+        def f(foo):
+            return foo
+
+        self.assertTrue(isinstance(safe_eval("f(Foo())", globals_dict={"f": f, "Foo": Foo}, mode="eval"), SafeWrapper))
```

Comment:
you added that line to check for 
```py
return SafeWrapper(__obj=obj, __type_checker=type_checker)
```
In `type_checker` right ?

But is the line really useful in the end ? (can we make a test where a check is not done because of that ? if no, we can just return `obj` instead of `SafeWrapper` in `type_checker`)